src/storage/txn/commands/check_txn_status.rs

```diff
@@ -1244,4 +1258,174 @@ pub mod tests {
             kvrpcpb::Op::Put,
         );
     }
+
+    #[test]
+    fn test_check_txn_status_resolving_primary_pessimistic_lock() {
+        let mut engine = TestEngineBuilder::new().build().unwrap();
+        let k1 = b"k1";
+        let v1 = b"v1";
+        let k2 = b"k2";
+        let v2 = b"v2";
+        let ts = TimeStamp::compose;
+
+        must_acquire_pessimistic_lock_with_ttl(&mut engine, k1, k1, ts(1, 0), ts(1, 0), 10);
+        must_acquire_pessimistic_lock_with_ttl(&mut engine, k2, k1, ts(1, 0), ts(1, 0), 10);
+        must_pessimistic_prewrite_put(
+            &mut engine,
+            k1,
+            v1,
+            k1,
+            ts(1, 0),
+            ts(1, 0),
+            DoPessimisticCheck,
+        );
+        must_pessimistic_prewrite_put(
+            &mut engine,
+            k2,
+            v2,
+            k1,
+            ts(1, 0),
+            ts(1, 0),
+            DoPessimisticCheck,
+        );
+        must_commit(&mut engine, k1, ts(1, 0), ts(2, 0));
+
+        // 1. Test resolve the stale pessimistic primary lock. Note the force lock
+        // could succeed only if there's no corresponding rollback record.
```

Comment:
When there is a rollback record, it seems it can still succeed if there is a newer record written by another transaction. This case is better to be covered too.